Feature: 스프링 시큐리티 및 Jwt 발급 기능 구현 #7

HuttTheJAVA opened the pull request:
# Feature: 스프링 시큐리티 및 Jwt 발급 기능 구현 #7

### ✅ PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [x] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### ✅ 개요
> 로그인에 사용할 Jwt 발급 기능 구현
스프링 시큐리티 필터 체인을 이용해 요청에 대해 인증, 인가

### ✅ 리뷰어가 꼭 봐줬으면 하는 부분
> 없습니다.

### ✅ ISSUE 번호
> 관련된 이슈 번호를 나열해주세요. 여러개 있다면 여러개 작성해주세요
- #7

### ✅ 참고 사항
> 1.주석을 추가하긴 했는데, 그래도 리뷰하시는데 이해 안되는 점 있으시면 바로 말씀해주세요!
2.WebSecurityConfig의 http 제한은 일단은 "/api/animal" 이거 하나입니다.
3.jwtProvider는 비밀키를 application.yml에서 주입받는데 아무 긴 문자열을 사용하면 됩니다. ex) spring.jwt.secret : "asdoqwjei123oj32~"

HuttTheJAVA committed:
feature: 스프링 시큐리티 및 Jwt 발급 기능 구현 #7
- 로그인에 사용할 Jwt 발급 기능 구현
- 스프링 시큐리티 필터 체인을 이용해 요청에 대해 인증, 인가

sngbmYu commented on src/main/java/com/jaefan/munpyspring/security/filter/handler/LoginSuccessHandler.java:
```diff
@@ -0,0 +1,65 @@
+	private void updateVisitAt(String email) {
+		Optional<User> optionalUser = userRepository.findByEmail(email);
+
+		User user = optionalUser.get();
```
혹시 Optional을 바로 get()으로 받으신 이유가 있을까요?
특별한 이유가 없다면 예외처리가 있으면 좋을 것 같습니다!

그리고 updateLastVisit()의 코드가 어떻게 되어 있는지는 모르겠지만,
값을 set 하는 메서드라면 dirty checking으로 자동 저장되니 save()는 필요 없을 것 같습니다!

HuttTheJAVA replied:
1. 이 코드가 인증이 성공 된 경우에만 동작하도록 구현된 코드라서, 바로 get하도록 작성한 것 같습니다. 그럼에도 말씀하신대로 예외에 대한 코드가 있는 것이 맞는 것 같습니다. (예외처리는 위의 JsonLoginFilter의 64번 줄에서 처리 됩니다.)
2. @Transactional이 붙은 메소드를 동일 클래스에서 직접 호출하면 트랜잭션 프록시 객체가 동작을 안 한다고 하기도 하고, 직접 Repository를 여기서 접근하는 게 이상한 것 같아서 updateLastVisit 처리는 UserService로 넘겼습니다.

말씀하신대로 나름 고쳐봤는데, 그럼에도 납득이 안되시는 점이 있으시면 바로 말씀해주세요!👍
UserService는 괜찮으시다면 다음 PR 때 같이 올리겠습니다! 혹시 코드가 궁금하실까봐 UserService의 updateLastVisit 코드 이미지를 같이 첨부하겠습니다.

<img width="601" alt="image" src="https://github.com/user-attachments/assets/f8eb7519-582f-4b22-98ff-74811d905f2b" />

sngbmYu reviewed (commented):
LoginSuccessHandler 클래스에 리뷰 남겼습니다.
확인하시고 답변주세요!

HuttTheJAVA committed:
fix: LoginSuccessHandler 코드 수정 #7
- email 유저 조회 시 EntityNotFoundException 예외 던짐 추가
- lastVisit 갱신 처리를 userRepository에서 userService로 위임 변경

Namgyu11 reviewed (approved):
고생 많으셨습니다!

LGTM 😎👍

sngbmYu reviewed (approved):
LGTM 👍
고생하셨습니다!
